In [1]:
import pandas as pd
import psycopg2 as pg

#### **1. GET THE DATA**

Import the data from PostgreSQL into dataframes

In [2]:
# Create the connection
engine = pg.connect("dbname='mundimoto' user='postgres' host='127.0.0.1' port='5432' password='sergio'")

In [3]:
# Get the brands
brands = pd.read_sql('select * from brands', con=engine)

C:\Users\spost\anaconda3\envs\mundimoto\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


In [4]:
# Get the versions
versions = pd.read_sql('select * from versions', con=engine)

C:\Users\spost\anaconda3\envs\mundimoto\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


#### **2. BASIC INSPECTION OF THE DATA**

**BRANDS table**

In [5]:
brands.head()

id     name
0   7      BMW
1  33  PEUGEOT
2  35  RENAULT
3  44   SUZUKI
4  51    ALFER

In [6]:
brands.shape

(307, 2)

There are 307 rows with ids and brands's names. Let's check if there are duplicates.

In [7]:
brands[brands.duplicated(subset=['name'], keep= False)]

id    name
0       7     BMW
3      44  SUZUKI
305  2415     BMW
306  2416  SUZUKI

The brands BMW and Suzuki appear two times

**VERSIONS table**

In [8]:
versions.head()

id                    name  brand_id  year  fuel  price
0   1   R 100 S   (1976-1979)         7  1976     2   3272
1   2   R 100 S   (1976-1979)         7  1977     2   6532
2   3  R 100 RT   (1978-1996)         7  1978     2   1724
3   4   R 100 S   (1976-1979)         7  1978     2   4599
4   5  R 100 RT   (1978-1996)         7  1979     2   5153

In [9]:
versions.shape

(39583, 6)

There are 39583. Let's check if there are duplicates.

In [10]:
versions[versions.duplicated(subset=['name', 'brand_id', 'year', 'fuel'], keep= False)]

id                                name  brand_id  year  fuel  price
872      873             R 1200 GS   (2017-2018)         7  2017     2   8769
873      874             R 1200 GS   (2017-2018)         7  2017     2   3036
874      875             R 1200 GS   (2017-2018)         7  2017     2   8923
910      911             R 1200 GS   (2017-2018)         7  2018     2   7305
911      912             R 1200 GS   (2017-2018)         7  2018     2   6222
912      913             R 1200 GS   (2017-2018)         7  2018     2   5329
7860    7861              FENIX 50   (1993-1997)        62  1993     2   2550
7861    7862              FENIX 50   (1993-1997)        62  1993     2   9123
7870    7871              FENIX 50   (1993-1997)        62  1994     2   9697
7871    7872              FENIX 50   (1993-1997)        62  1994     2   1684
7881    7882              FENIX 50   (1993-1997)        62  1995     2   8732
7882    7883              FENIX 50   (1993-1997)        62  1995     2  10185
7891    7892              FENIX 50   (1993-1997)        62  1996     2   3941
7892    7893              FENIX 50   (1993-1997)        62  1996     2   7696
7900    7901              FENIX 50   (1993-1997)        62  1997     2   7213
7901    7902              FENIX 50   (1993-1997)        62  1997     2   1855
16850  16854    GT 125 R Pro Comet   (2016-2018)        72  2016     2   8508
16851  16855    GT 125 R Pro Comet   (2016-2018)        72  2016     2   8527
16867  16871    GT 125 R Pro Comet   (2016-2018)        72  2017     2  10942
16868  16872    GT 125 R Pro Comet   (2016-2018)        72  2017     2   3308
16881  16885    GT 125 R Pro Comet   (2016-2018)        72  2018     2   9082
16882  16886    GT 125 R Pro Comet   (2016-2018)        72  2018     2   7690
22579  22583              STRADALE   (2015-2018)        84  2015     2   5236
22581  22585              STRADALE   (2015-2018)        84  2015     2   5435
22626  22630              STRADALE   (2015-2018)        84  2016     2   2618
22627  22631              STRADALE   (2015-2018)        84  2016     2   5718
22644  22648              STRADALE   (2015-2018)        84  2017     2   8437
22645  22649              STRADALE   (2015-2018)        84  2017     2   2650
22662  22666              STRADALE   (2015-2018)        84  2018     2   4771
22663  22667              STRADALE   (2015-2018)        84  2018     2   5640
30364  30368       XJ6 Diversion S   (2009-2010)       101  2009     2   6838
30365  30369       XJ6 Diversion S   (2009-2010)       101  2009     2   8182
30468  30472       XJ6 Diversion S   (2009-2010)       101  2010     2   5154
30469  30473       XJ6 Diversion S   (2009-2010)       101  2010     2   2306
30621  30625  TZR 50th Anniversary   (2012-2013)       101  2012     2   4206
30622  30626  TZR 50th Anniversary   (2012-2013)       101  2012     2   6645
30721  30725  TZR 50th Anniversary   (2012-2013)       101  2013     2   5029
30722  30726  TZR 50th Anniversary   (2012-2013)       101  2013     2   1732
31009  31013          SUPER TÉNÉRÉ E   (2016)       101  2016     2   8865
31010  31014          SUPER TÉNÉRÉ E   (2016)       101  2016     2   1873
31958  31962    Desert Phantom 250   (1974-1978)       197  1974     2   8635
31959  31963    Desert Phantom 250   (1974-1978)       197  1974     2   4990
31965  31969    Desert Phantom 250   (1974-1978)       197  1975     2   7482
31966  31970    Desert Phantom 250   (1974-1978)       197  1975     2   4464
31975  31979    Desert Phantom 250   (1974-1978)       197  1976     2   9534
31976  31980    Desert Phantom 250   (1974-1978)       197  1976     2   7548
31987  31991    Desert Phantom 250   (1974-1978)       197  1977     2   4310
31988  31992    Desert Phantom 250   (1974-1978)       197  1977     2   6571
32000  32004    Desert Phantom 250   (1974-1978)       197  1978     2   7907
32001  32005    Desert Phantom 250   (1974-1978)       197  1978     2   3060
33585  33589         450 SEF-R  

#### **3.  DATA CLEANING**

According to Mundimoto, for the Hackathon this prices are generated randomly, so they told me to simple remove the duplicates and keep one without taking care of the price. I do that:

In [11]:
versions.drop_duplicates(subset=['name', 'brand_id', 'year', 'fuel'], inplace = True)

Now let's see if the duplicated brand's ids in brands tables are used both in versions table

In [12]:
# Get a list with the duplicated ids
duplicated_brand_ids = brands[brands.duplicated(subset=['name'], keep= False)]["id"].tolist()
duplicated_brand_ids

[7, 44, 2415, 2416]

In [13]:
# Now check if a same brand appears with different brand_ids in versions
versions[versions['brand_id'].isin(duplicated_brand_ids)]

id                                  name  brand_id  year  fuel  \
0          1                 R 100 S   (1976-1979)         7  1976     2   
1          2                 R 100 S   (1976-1979)         7  1977     2   
2          3                R 100 RT   (1978-1996)         7  1978     2   
3          4                 R 100 S   (1976-1979)         7  1978     2   
4          5                R 100 RT   (1978-1996)         7  1979     2   
...      ...                                   ...       ...   ...   ...   
4342    4343         V-Strom 250   (modelo actual)        44  2021     2   
4343    4344     V-Strom 650 ABS   (modelo actual)        44  2021     2   
4344    4345  V-Strom 650 XT ABS   (modelo actual)        44  2021     2   
39581  39585                  F 800 GS 30 YEARS GS      2415  1985     2   
39582  39586                  F 800 GS 30 YEARS GS      2416  2003     1   

       price  
0       3272  
1       6532  
2       1724  
3       4599  
4       5153  
...      ...  
4342    1723  
4343    6846  
4344    2285  
39581   5649  
39582   5611  

[2969 rows x 6 columns]

They do. We will unified the ones pointing to the same brand_id. For that we will create a mapping

In [14]:
di = pd.merge(brands, brands, how="inner", on='name')
di = di.drop(di[di.id_x >= di.id_y].index)
di = di.set_index('id_x').to_dict()['id_y']
di

{7: 2415, 44: 2416}

Finally, map the corresponding brand_ids in the versions table and in brands table and remove duplicates.

In [15]:
versions = versions.replace({"brand_id": di})

In [16]:
brands = brands.replace({"id": di}).drop_duplicates()

Make sure it worked, the brands table should not have duplicated names anymore

In [17]:
brands[brands.duplicated(subset=['name'], keep= False)]

Empty DataFrame
Columns: [id, name]
Index: []

Finally, let's join the brands and versions table to have a consolidated table

In [18]:
motorcycles = pd.merge(brands,versions,left_on='id',right_on='brand_id')[['id_x', 'name_x', 'name_y', 'year', 'fuel', 'price' ]]
motorcycles.rename(columns = {'id_x':'id', 'name_x':'brand', 'name_y':'model'}, inplace = True)
motorcycles

id   brand                           model  year  fuel  price
0      2415     BMW           R 100 S   (1976-1979)  1976     2   3272
1      2415     BMW           R 100 S   (1976-1979)  1977     2   6532
2      2415     BMW          R 100 RT   (1978-1996)  1978     2   1724
3      2415     BMW           R 100 S   (1976-1979)  1978     2   4599
4      2415     BMW          R 100 RT   (1978-1996)  1979     2   5153
...     ...     ...                             ...   ...   ...    ...
39548  2412   URBET  Gadiro E 125   (modelo actual)  2021     3   7818
39549  2412   URBET          Nura   (modelo actual)  2021     3   8722
39550  2412   URBET        Riazor   (modelo actual)  2021     3   9483
39551  2414  BENSOM        Moscú   (modelo actual)  2021     3   7589
39552  2414  BENSOM         Tokyo   (modelo actual)  2021     3   7208

[39553 rows x 6 columns]

Finally, export the cleaned data into CSV

In [19]:
motorcycles.to_csv (r'.\motorcycles.csv', index = False, header=True)
